### ISSUE TITLE
Failed to compile Giraffe

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtimelab#1357]
I try to compile this app https://github.com/giraffe-fsharp/Giraffe/tree/master/samples/EndpointRoutingApp/EndpointRoutingApp

```
  Unhandled exception. ILCompiler.CodeGenerationFailedException: Code generation failed for method '[Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.TaskAwaiter`1<System.__Canon>,Microsoft.FSharp.Core.FSharpResult`2<System.__Canon,System.__Canon>>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<FSharpResult`2<__Canon,__Canon>>>(ContinuationStateMachine`1<FSharpResult`2<__Canon,__Canon>>&)'
   ---> System.InvalidOperationException: [Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.TaskAwaiter`1<System.__Canon>,Microsoft.FSharp.Core.FSharpResult`2<System.__Canon,System.__Canon>>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<FSharpResult`2<__Canon,__Canon>>>(ContinuationStateMachine`1<FSharpResult`2<__Canon,__Canon>>&): MethodDictionary: [Ply]Ply.Internal+IAwaitingMachine.AwaitUnsafeOnCompleted<TaskAwaiter`1<__Canon>>(TaskAwaiter`1<__Canon>&)
     at ILCompiler.DependencyAnalysis.PrecomputedDictionaryLayoutNode.GetSlotForEntry(GenericLookupResult entry) in ILCompiler.Compiler.dll:token 0x60009d9+0x27
     at ILCompiler.Compilation.ComputeGenericLookup(MethodDesc contextMethod, ReadyToRunHelperId lookupKind, Object targetOfLookup) in ILCompiler.Compiler.dll:token 0x600031a+0xdd
     at Internal.JitInterface.CorInfoImpl.ComputeLookup(CORINFO_RESOLVED_TOKEN& pResolvedToken, Object entity, ReadyToRunHelperId helperId, CORINFO_LOOKUP& lookup) in ILCompiler.RyuJit.dll:token 0x6000045+0x55
     at Internal.JitInterface.CorInfoImpl.embedGenericHandle(CORINFO_RESOLVED_TOKEN& pResolvedToken, Boolean fEmbedParent, CORINFO_GENERICHANDLE_RESULT& pResult) in ILCompiler.RyuJit.dll:token 0x6000058+0x165
     at Internal.JitInterface.CorInfoImpl._embedGenericHandle(IntPtr thisHandle, IntPtr* ppException, CORINFO_RESOLVED_TOKEN* pResolvedToken, Byte fEmbedParent, CORINFO_GENERICHANDLE_RESULT* pResult) in ILCompiler.RyuJit.dll:token 0x60000f0+0x7
     --- End of inner exception stack trace ---
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x600012a+0xae
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodCodeNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x6000043+0x28
     at ILCompiler.RyuJitCompilation.CompileSingleMethod(CorInfoImpl corInfo, MethodCodeNode methodCodeNodeNeedingCode) in ILCompiler.RyuJit.dll:token 0x600022f+0x7
     at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in System.Private.CoreLib.dll:token 0x6002b1d+0x14
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002af8+0xa0
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002bd2+0x6c
     at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600278e+0xe
  Unhandled exception. Unhandled exception. Unhandled exception. ILCompiler.CodeGenerationFailedException: Code generation failed for method '[Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.ValueTaskAwaiter`1<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&)'
   ---> System.InvalidOperationException: [Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.ValueTaskAwaiter`1<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&): MethodDictionary: [Ply]Ply.Internal+IAwaitingMachine.AwaitUnsafeOnCompleted<ValueTaskAwaiter`1<__Canon>>(ValueTaskAwaiter`1<__Canon>&)
     at ILCompiler.DependencyAnalysis.PrecomputedDictionaryLayoutNode.GetSlotForEntry(GenericLookupResult entry) in ILCompiler.Compiler.dll:token 0x60009d9+0x27
     at ILCompiler.Compilation.ComputeGenericLookup(MethodDesc contextMethod, ReadyToRunHelperId lookupKind, Object targetOfLookup) in ILCompiler.Compiler.dll:token 0x600031a+0xdd
     at Internal.JitInterface.CorInfoImpl.ComputeLookup(CORINFO_RESOLVED_TOKEN& pResolvedToken, Object entity, ReadyToRunHelperId helperId, CORINFO_LOOKUP& lookup) in ILCompiler.RyuJit.dll:token 0x6000045+0x55
     at Internal.JitInterface.CorInfoImpl.embedGenericHandle(CORINFO_RESOLVED_TOKEN& pResolvedToken, Boolean fEmbedParent, CORINFO_GENERICHANDLE_RESULT& pResult) in ILCompiler.RyuJit.dll:token 0x6000058+0x165
     at Internal.JitInterface.CorInfoImpl._embedGenericHandle(IntPtr thisHandle, IntPtr* ppException, CORINFO_RESOLVED_TOKEN* pResolvedToken, Byte fEmbedParent, CORINFO_GENERICHANDLE_RESULT* pResult) in ILCompiler.RyuJit.dll:token 0x60000f0+0x7
     --- End of inner exception stack trace ---
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x600012a+0xae
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodCodeNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x6000043+0x28
     at ILCompiler.RyuJitCompilation.CompileSingleMethod(CorInfoImpl corInfo, MethodCodeNode methodCodeNodeNeedingCode) in ILCompiler.RyuJit.dll:token 0x600022f+0x7
     at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in System.Private.CoreLib.dll:token 0x6002b1d+0x14
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002af8+0xa0
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002bd2+0x6c
     at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600278e+0xe
  ILCompiler.CodeGenerationFailedException: Code generation failed for method '[Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&)'
   ---> System.InvalidOperationException: [Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1+ConfiguredValueTaskAwaiter<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&): MethodDictionary: [Ply]Ply.Internal+IAwaitingMachine.AwaitUnsafeOnCompleted<ConfiguredValueTaskAwaiter<__Canon>>(ConfiguredValueTaskAwaiter<__Canon>&)
     at ILCompiler.DependencyAnalysis.PrecomputedDictionaryLayoutNode.GetSlotForEntry(GenericLookupResult entry) in ILCompiler.Compiler.dll:token 0x60009d9+0x27
     at ILCompiler.Compilation.ComputeGenericLookup(MethodDesc contextMethod, ReadyToRunHelperId lookupKind, Object targetOfLookup) in ILCompiler.Compiler.dll:token 0x600031a+0xdd
     at Internal.JitInterface.CorInfoImpl.ComputeLookup(CORINFO_RESOLVED_TOKEN& pResolvedToken, Object entity, ReadyToRunHelperId helperId, CORINFO_LOOKUP& lookup) in ILCompiler.RyuJit.dll:token 0x6000045+0x55
     at Internal.JitInterface.CorInfoImpl.embedGenericHandle(CORINFO_RESOLVED_TOKEN& pResolvedToken, Boolean fEmbedParent, CORINFO_GENERICHANDLE_RESULT& pResult) in ILCompiler.RyuJit.dll:token 0x6000058+0x165
     at Internal.JitInterface.CorInfoImpl._embedGenericHandle(IntPtr thisHandle, IntPtr* ppException, CORINFO_RESOLVED_TOKEN* pResolvedToken, Byte fEmbedParent, CORINFO_GENERICHANDLE_RESULT* pResult) in ILCompiler.RyuJit.dll:token 0x60000f0+0x7
     --- End of inner exception stack trace ---
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x600012a+0xae
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodCodeNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x6000043+0x28
     at ILCompiler.RyuJitCompilation.CompileSingleMethod(CorInfoImpl corInfo, MethodCodeNode methodCodeNodeNeedingCode) in ILCompiler.RyuJit.dll:token 0x600022f+0x7
     at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in System.Private.CoreLib.dll:token 0x6002b1d+0x14
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002af8+0xa0
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002bd2+0x6c
     at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600278e+0xe
  Unhandled exception. ILCompiler.CodeGenerationFailedException: Code generation failed for method '[Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.TaskAwaiter`1<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&)'
   ---> System.InvalidOperationException: [Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.TaskAwaiter`1<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&): MethodDictionary: [Ply]Ply.Internal+IAwaitingMachine.AwaitUnsafeOnCompleted<TaskAwaiter`1<__Canon>>(TaskAwaiter`1<__Canon>&)
     at ILCompiler.DependencyAnalysis.PrecomputedDictionaryLayoutNode.GetSlotForEntry(GenericLookupResult entry) in ILCompiler.Compiler.dll:token 0x60009d9+0x27
     at ILCompiler.Compilation.ComputeGenericLookup(MethodDesc contextMethod, ReadyToRunHelperId lookupKind, Object targetOfLookup) in ILCompiler.Compiler.dll:token 0x600031a+0xdd
     at Internal.JitInterface.CorInfoImpl.ComputeLookup(CORINFO_RESOLVED_TOKEN& pResolvedToken, Object entity, ReadyToRunHelperId helperId, CORINFO_LOOKUP& lookup) in ILCompiler.RyuJit.dll:token 0x6000045+0x55
     at Internal.JitInterface.CorInfoImpl.embedGenericHandle(CORINFO_RESOLVED_TOKEN& pResolvedToken, Boolean fEmbedParent, CORINFO_GENERICHANDLE_RESULT& pResult) in ILCompiler.RyuJit.dll:token 0x6000058+0x165
     at Internal.JitInterface.CorInfoImpl._embedGenericHandle(IntPtr thisHandle, IntPtr* ppException, CORINFO_RESOLVED_TOKEN* pResolvedToken, Byte fEmbedParent, CORINFO_GENERICHANDLE_RESULT* pResult) in ILCompiler.RyuJit.dll:token 0x60000f0+0x7
     --- End of inner exception stack trace ---
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x600012a+0xae
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodCodeNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x6000043+0x28
     at ILCompiler.RyuJitCompilation.CompileSingleMethod(CorInfoImpl corInfo, MethodCodeNode methodCodeNodeNeedingCode) in ILCompiler.RyuJit.dll:token 0x600022f+0x7
     at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in System.Private.CoreLib.dll:token 0x6002b1d+0x14
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002af8+0xa0
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002bd2+0x6c
     at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600278e+0xe
  Unhandled exception. ILCompiler.CodeGenerationFailedException: Code generation failed for method '[Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&)'
   ---> System.InvalidOperationException: [Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter<System.__Canon>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&): MethodDictionary: [Ply]Ply.Internal+IAwaitingMachine.AwaitUnsafeOnCompleted<ConfiguredTaskAwaiter<__Canon>>(ConfiguredTaskAwaiter<__Canon>&)
     at ILCompiler.DependencyAnalysis.PrecomputedDictionaryLayoutNode.GetSlotForEntry(GenericLookupResult entry) in ILCompiler.Compiler.dll:token 0x60009d9+0x27
     at ILCompiler.Compilation.ComputeGenericLookup(MethodDesc contextMethod, ReadyToRunHelperId lookupKind, Object targetOfLookup) in ILCompiler.Compiler.dll:token 0x600031a+0xdd
     at Internal.JitInterface.CorInfoImpl.ComputeLookup(CORINFO_RESOLVED_TOKEN& pResolvedToken, Object entity, ReadyToRunHelperId helperId, CORINFO_LOOKUP& lookup) in ILCompiler.RyuJit.dll:token 0x6000045+0x55
     at Internal.JitInterface.CorInfoImpl.embedGenericHandle(CORINFO_RESOLVED_TOKEN& pResolvedToken, Boolean fEmbedParent, CORINFO_GENERICHANDLE_RESULT& pResult) in ILCompiler.RyuJit.dll:token 0x6000058+0x165
     at Internal.JitInterface.CorInfoImpl._embedGenericHandle(IntPtr thisHandle, IntPtr* ppException, CORINFO_RESOLVED_TOKEN* pResolvedToken, Byte fEmbedParent, CORINFO_GENERICHANDLE_RESULT* pResult) in ILCompiler.RyuJit.dll:token 0x60000f0+0x7
     --- End of inner exception stack trace ---
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x600012a+0xae
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodCodeNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x6000043+0x28
     at ILCompiler.RyuJitCompilation.CompileSingleMethod(CorInfoImpl corInfo, MethodCodeNode methodCodeNodeNeedingCode) in ILCompiler.RyuJit.dll:token 0x600022f+0x7
     at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in System.Private.CoreLib.dll:token 0x6002b1d+0x14
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002af8+0xa0
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002bd2+0x6c
     at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600278e+0xe
  ILCompiler.CodeGenerationFailedException: Code generation failed for method '[Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.TaskAwaiter`1<Microsoft.FSharp.Core.FSharpResult`2<System.__Canon,System.__Canon>>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&)'
   ---> System.InvalidOperationException: [Ply]Ply.TplPrimitives+TplAwaitable`2<System.Runtime.CompilerServices.TaskAwaiter`1<Microsoft.FSharp.Core.FSharpResult`2<System.__Canon,System.__Canon>>,System.__Canon>.Ply-Internal-IAwaitable`1-Await<ContinuationStateMachine`1<__Canon>>(ContinuationStateMachine`1<__Canon>&): MethodDictionary: [Ply]Ply.Internal+IAwaitingMachine.AwaitUnsafeOnCompleted<TaskAwaiter`1<FSharpResult`2<__Canon,__Canon>>>(TaskAwaiter`1<FSharpResult`2<__Canon,__Canon>>&)
     at ILCompiler.DependencyAnalysis.PrecomputedDictionaryLayoutNode.GetSlotForEntry(GenericLookupResult entry) in ILCompiler.Compiler.dll:token 0x60009d9+0x27
     at ILCompiler.Compilation.ComputeGenericLookup(MethodDesc contextMethod, ReadyToRunHelperId lookupKind, Object targetOfLookup) in ILCompiler.Compiler.dll:token 0x600031a+0xdd
     at Internal.JitInterface.CorInfoImpl.ComputeLookup(CORINFO_RESOLVED_TOKEN& pResolvedToken, Object entity, ReadyToRunHelperId helperId, CORINFO_LOOKUP& lookup) in ILCompiler.RyuJit.dll:token 0x6000045+0x55
     at Internal.JitInterface.CorInfoImpl.embedGenericHandle(CORINFO_RESOLVED_TOKEN& pResolvedToken, Boolean fEmbedParent, CORINFO_GENERICHANDLE_RESULT& pResult) in ILCompiler.RyuJit.dll:token 0x6000058+0x165
     at Internal.JitInterface.CorInfoImpl._embedGenericHandle(IntPtr thisHandle, IntPtr* ppException, CORINFO_RESOLVED_TOKEN* pResolvedToken, Byte fEmbedParent, CORINFO_GENERICHANDLE_RESULT* pResult) in ILCompiler.RyuJit.dll:token 0x60000f0+0x7
     --- End of inner exception stack trace ---
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x600012a+0xae
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodCodeNode methodCodeNodeNeedingCode, MethodIL methodIL) in ILCompiler.RyuJit.dll:token 0x6000043+0x28
     at ILCompiler.RyuJitCompilation.CompileSingleMethod(CorInfoImpl corInfo, MethodCodeNode methodCodeNodeNeedingCode) in ILCompiler.RyuJit.dll:token 0x600022f+0x7
     at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in System.Private.CoreLib.dll:token 0x6002b1d+0x14
     at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002af8+0xa0
     at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002bd2+0x6c
     at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600278e+0xe
```

**Comment by jkotas** [comment]
Scanner bug. cc @MichalStrehovsky

**Comment by MichalStrehovsky** [comment]
Simple repro:

```csharp
using System;
using System.Runtime.CompilerServices;

Gen.DoInGen<Foo<object>, object, string>(default);

class Gen
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void DoInGen<U, V, W>(U t) where U : IFoo<V>
    {
        t.DoMe<W>();
    }
}

interface IFoo<T>
{
    void DoMe<U>();
}

struct Foo<T> : IFoo<T>
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public void DoMe<U>() {}
}
```

RyuJIT is asking for generic dictionary of the interface method, not of the implementation (in the constrained call).

This might require a JitInterface change. I'll need to investigate how this is dealt with on the CoreCLR/crossgen side.

**Comment by MichalStrehovsky** [comment]
Should be fixed in #1422.
